Add docs for image sizes (#41)
humanmade/altis-media: docs/image-sizes.md

# Defining Image Sizes

Altis provides some default image sizes, that you can use to render images in the theme.

- Thumbnail
- Medium
- Large
- Post Thumbnail
- Full

They have default dimensions, but you can override these either with code or in the admin, under [Settings > Media](internal://admin/options-media.php).

![Media Settings](./assets/media-settings.png)

A lot of times, you'll find that you need custom image sizes for different contexts, depending on your theme's design.

You can define custom image sizes with the `add_image_size()` function.

This function accepts the following parameters:

```php
 * @param string     $name   Image size identifier.
 * @param int        $width  Optional. Image width in pixels. Default 0.
 * @param int        $height Optional. Image height in pixels. Default 0.
 * @param bool|array $crop   Optional. Whether to crop images to specified width and height or resize.
```
The custom image sizes should be declared in the callback function to the `after_setup_theme` action.

Example:
```php
add_action( 'after_setup_theme', 'theme_setup' );
function theme_setup() {
	// Set the image size by resizing the image while maintaining the aspect ratio:

> 	add_image_size( 'custom-size', 220, 180 ); // Image constrained to 220 pixels wide by 180 pixels tall

Reviewer: ```suggestion
	add_image_size( 'constrained-thumbnail', 220, 180 ); // Image constrained to 220 pixels wide by 180 pixels tall
```

	
	// Set the image size by cropping the image (not showing part of it):

> 	add_image_size( 'custom-size', 220, 180, true ); // Exactly 220 pixels wide by 180 pixels tall

Reviewer: ```suggestion
	add_image_size( 'cropped-thumbnail', 220, 180, true ); // Exactly 220 pixels wide by 180 pixels tall
```

	
	// Set the image size by cropping the image and defining a crop position:

> 	add_image_size( 'custom-size', 220, 220, [ 'left', 'top' ] ); // Hard crop left top

Reviewer: ```suggestion
	add_image_size( 'top-left-crop', 220, 220, [ 'left', 'top' ] ); // Hard crop left top
```

}
```

Custom images sizes do not show up in the editor by default when editing image blocks. See the [Hooks and Filters section](#hooks-and-filters) below to find out how to do this.

### Cropping

Cropping behavior for the image size is dependent on the value of `$crop`:

> 1. If false (default), images will be scaled, not cropped.

Reviewer: ```suggestion
1. If `false` (default), images will be scaled, not cropped.
```

2. If an array in the form of `[ x_crop_position, y_crop_position ]`:
    - `x_crop_position` accepts 'left' 'center', or 'right'.
    - `y_crop_position` accepts 'top', 'center', or 'bottom'.
    Images will be cropped to the specified dimensions within the defined crop area.

> 3. If true, images will be cropped to the specified dimensions using center positions.

Reviewer: ```suggestion
3. If `true`, images will be cropped to the specified dimensions using smart cropping by default, otherwise falling back to center cropping.
```

 
You can find more info in the [WordPress developer guide](https://developer.wordpress.org/themes/functionality/featured-images-post-thumbnails/#add-custom-featured-image-sizes)

## Using The Custom Sizes

Once you've defined custom image sizes there are different ways you can use them.

> You can display a post thumbnail:

Reviewer: ```suggestion
You can display a post's featured image at a given size:
```

```php
// Output the post thumbnail image tag:

> the_post_thumbnail( 'custom-size` );

Reviewer: ```suggestion
the_post_thumbnail( 'custom-size' );
```

```

> You can display any uploaded image by its attachment ID:

Reviewer: ```suggestion
You can also display any uploaded image by its attachment ID:
```

```php
$src = wp_get_attachment_image_src( $attachment_id , 'custom-size' );
$img_url = $src[0];
echo ( ! empty( $img_url ) ) ? $img_url : '';
```

## Hooks And Filters

### `image_size_names_choose`

This filter allows you to make your custom image sizes available for selection in the admin.

> Runs before rendering the list of available image sizes in the sidebar of the media edit screen. Used in conjuction with `add_image_size` 

Reviewer: ```suggestion
Runs before rendering the list of available image sizes in the sidebar of the media edit screen. Used in conjuction with `add_image_size()` 
```

```php
add_filter( 'image_size_names_choose', function ( array $sizes ) : array {
	$sizes['category-thumb'] = __( 'Category Thumb' );
	return $sizes;

> });

Reviewer: ```suggestion
} );
```

# Images Sizes

You can define additional image thumbnail sizes with `add_image_size`.

> The custom thumbnail sizes should be declared in the callback function to the `after_setup_theme` action.

Reviewer: You should document the function `add_image_size()` and the parameters it takes before this. Right now the example below doesn't call special attention to the `add_image_size()` so it's not as scannable as it could be.

Try and think that people reading this might have no prior WordPress or Altis knowledge.

Example:
```php
add_action( 'after_setup_theme', __NAMESPACE__ . '\\theme_setup' );
function theme_setup() {

>     add_image_size( 'category-thumb', 300, 9999 ); // 300 pixels wide (and unlimited height)

Reviewer: We need more examples here of all the different types of sizes you can generate, so that includes cropped images. This is a bit too light touch as it is I think, I shouldn't have to visit the WP links below to understand what this does and how to use it.

}
```

> You can find more info at the [WordPress Developer Reference](https://developer.wordpress.org/reference/functions/add_image_size/) and [WordPress developer guide](https://developer.wordpress.org/themes/functionality/featured-images-post-thumbnails/#add-custom-featured-image-sizes)

Reviewer: Once the `add_image_size()` function is documented here let's ditch the link to the WP developer reference. I'm a bit torn on whether to include the WP developer guide link but I guess it has a bunch of extra detail.

> ## Hooks and filters

Reviewer: ```suggestion
## Hooks and Filters
```

You can add custom image sizes by hooking in on this filter.

### `image_size_names_choose`

Runs before rendering the list of available thumbnail sizes in the sidebar of the media edit screen. Used in conjuction with `add_image_size` 

```php

>     add_filter( 'image_size_names_choose', function ( array $sizes ) {

Reviewer: No need to indent this. Seems to be using spaces for indentation too, we use tabs in all the other code examples so best to stay consistent.

>     add_filter( 'image_size_names_choose', function ( array $sizes ) {

Reviewer: ```suggestion
add_filter( 'image_size_names_choose', function ( array $sizes ) : array {
```
Can add the return type hint too

        $sizes['category-thumb'] = __( 'Category Thumb' );
        return $sizes;
    });
```

You can find the [full documentation for this filter here](https://developer.wordpress.org/reference/hooks/image_size_names_choose/).

> ## Dynamically defined with a size array

Reviewer: We use title casing in all the docs so should be 
```suggestion
## Dynamically Defined With a Size Array
```

There are scenarios where you might want to define the thumbnail size via an array:

> `$thumb_url = wp_get_attachment_image_src( $attachment_id, [ 900, 450 ], true );`

Reviewer: Code samples that are on their own line should be in a proper code block with the 3 backticks and language identifier eg. ```php

> 

Reviewer: It's important to note as well that you can't get a cropped image using this method. I'd like to make that possible though so raising an issue on tachyon-plugin :)

> This will generate a thumbnail on the fly with the dimensions defined by an array of width and height [w, h] values in pixels

Reviewer: Full stop at the end of the sentence.

- Thumbnail
- Medium
- Large
- Post Thumbnail
- Full

They have default dimensions, but you can override these either with code or in the Admin, under Settings > Media.

![Media Settings](https://user-images.githubusercontent.com/30460/67079140-8b846000-f18a-11e9-8387-038b594f19aa.png)

A lot of times, you'll find that you need additional image sizes for different contexts, depending on your theme's design.

You can define additional image sizes with  the`add_image_size` function.

This function accepts the following parameters:

```php
 * @param string     $name   Image size identifier.
 * @param int        $width  Optional. Image width in pixels. Default 0.
 * @param int        $height Optional. Image height in pixels. Default 0.
 * @param bool|array $crop   Optional. Whether to crop images to specified width and height or resize.
```
The custom image sizes should be declared in the callback function to the `after_setup_theme` action.

Example:
```php
add_action( 'after_setup_theme', __NAMESPACE__ . '\\theme_setup' );
function theme_setup() {
	// Set the image size by resizing the image proportionally (without distorting it):

> 	add_image_size( 'custom-size', 220, 180 ); // 220 pixels wide by 180 pixels tall, soft proportional crop mode

Reviewer: ```suggestion
	add_image_size( 'custom-size', 220, 180 ); // Image constrained to 220 pixels wide by 180 pixels tall
```
My feeling here is that we're not actually doing any cropping in this instance so "soft crop" doesn't really say anything.

	
	// Set the image size by cropping the image (not showing part of it):

> 	add_image_size( 'custom-size', 220, 180, true ); // 220 pixels wide by 180 pixels tall, hard crop mode

Reviewer: ```suggestion
	add_image_size( 'custom-size', 220, 180, true ); // Exactly 220 pixels wide by 180 pixels tall
```

	
	// Set the image size by cropping the image and defining a crop position:

> 	add_image_size( 'custom-size', 220, 220, array( 'left', 'top' ) ); // Hard crop left top

Reviewer: I've generally aimed to make these examples as real-world as possible - the main issue I see here is that the size name is `custom-size` for all 3 - in reality each one would overwrite the last.

}
```

> This will *not* add a UI for these named sizes under Settings > Media.

Reviewer: So these image sizes will available to set custom crops for, it's just in the editor you don't get to choose the sizes when writing content. Can you make this a bit longer, something like
```suggestion
Custom images sizes do not show up in the editor by default when editing image blocks. See the [Hooks and Filters section](#hooks-and-filters) below to find out how to do this.
```

### Cropping

> Cropping behavior for the image size is dependent on the value of $crop:

Reviewer: ```suggestion
Cropping behavior for the image size is dependent on the value of `$crop`:
```
Code should always be marked up as such

1. If false (default), images will be scaled, not cropped.

> 2. If an array in the form of array( x_crop_position, y_crop_position ):

Reviewer: Again mark up code as code
```suggestion
2. If an array in the form of `[ x_crop_position, y_crop_position ]`:
```

>     - x_crop_position accepts 'left' 'center', or 'right'.

Reviewer: ```suggestion
    - `x_crop_position` accepts 'left' 'center', or 'right'.
```

>     - y_crop_position accepts 'top', 'center', or 'bottom'.

Reviewer: ```suggestion
    - `y_crop_position` accepts 'top', 'center', or 'bottom'.
```

    Images will be cropped to the specified dimensions within the defined crop area.

>  3. If true, images will be cropped to the specified dimensions using center positions.

Reviewer: ```suggestion
3. If true, images will be cropped to the specified dimensions using center positions.
```
rando space snuck in!

Reviewer: In Altis smart cropping applies by default so we can update it to mention that.

 
You can find more info in the [WordPress developer guide](https://developer.wordpress.org/themes/functionality/featured-images-post-thumbnails/#add-custom-featured-image-sizes)

## Using The Custom Sizes

> Once you've defined custom image sizes, there are different ways you can use them.

Reviewer: Bonus comma
```suggestion
Once you've defined custom image sizes there are different ways you can use them.
```

You can display a post thumbnail: `the_post_thumbnail( 'custom-size` );`

> You can display any uploaded image by its attachment ID: `wp_get_attachment_image_src( 1 , 'custom-size' );`

Reviewer: I'd consider making this into a code example. Personal preference but I'm less keen on inline code examples. So something like:

```php
// Output the post thumbnail image tag:
the_post_thumbnail( 'custom-size' );
/// etc...
```

## Hooks And Filters

### `image_size_names_choose`

This filter allows you to make your custom image sizes available for selection in the admin.

Runs before rendering the list of available image sizes in the sidebar of the media edit screen. Used in conjuction with `add_image_size` 

```php
add_filter( 'image_size_names_choose', function ( array $sizes ) : array {
	$sizes['category-thumb'] = __( 'Category Thumb' );
	return $sizes;
});
```

> ![Media Editor](https://user-images.githubusercontent.com/30460/67079152-93dc9b00-f18a-11e9-9797-5075210affad.png)

Reviewer: Again add the image to an `assets` folder and reference from there

You can find the [full documentation for this filter here](https://developer.wordpress.org/reference/hooks/image_size_names_choose/).

## Dynamically Defined With A Size Array

> There are scenarios where you might want to define the image size via an array:

Reviewer: Thinking about this some more we can actually define the general scenario for this:
```suggestion
In cases where a specific image size might only be used once you can define the image size via an array of width and height values in pixels:
```

```php
$thumb_url = wp_get_attachment_image_src( $attachment_id, [ 900, 450 ], true );`
```

> This will generate an image on the fly with the dimensions defined by an array of width and height [w, h] values in pixels.

Reviewer: ```suggestion
```
We can scrub this line now, made redundant by putting more information above.

> This function doesn't allow setting the crop values though.

Reviewer: This sentence is a bit lost on its own like this, and the use of "though" is a bit too conversational I think. usually for a callout we've been following a pattern like this:

```suggestion
**NOTE**: This method of defining dimensions currently does not support cropping.
```
